### ISSUE TITLE
Reconnect Midi

### ISSUE THREAD

**Issue by taurojo** [naudio/NAudio#445]
Hi,

I need to detect when the midi is disconnected.
In your example(Midi IN), it does not detect connection or disconnection. It does not send any warning or detect if it has been disconnected.

Or if it is possible to reconnect automatically with the midi device again.

it's possible?
Thanks

**Comment by markheath** [comment]
I don't think that;s something you can know I'm afraid. If it was USB midi then Windows might know that the USB device had disconnected. I think the new MIDI APIs in Windows 10 might have device disconnected capabilities. But as far as I know there's nothing in NAudio that can do that at the moment

**Comment by taurojo** [comment]
Thanks a lot,
I go to revise:
[https://github.com/Microsoft/Windows-universal-samples/tree/master/Samples/MIDI](url)

**Comment by MovGP0** [comment]
It's a missing feature. On UWP you can add a reference to `Microsoft General MIDI DLS for Universal Windows Apps` found in `Add Reference...` ⇒ `Universal Windows` ⇒ `Extensions`.

You can then use it like this to set up a watcher: 

```csharp
using Windows.Devices.Enumeration; 
using System.Reactive.Disposables;
using Windows.Devices.Midi;

    private void StartWatchingOutputDevices(CompositeDisposable d)
    {
           var watcher = DeviceInformation.CreateWatcher(MitiOutPort.GetDeviceSelector());

            watcher.Added += OnMidiOutputDeviceAdded;
            watcher.Removed += OnMidiOutputDeviceRemoved;
            watcher.Updated += OnMidiOutputDeviceUpdated;
            watcher.EnumerationCompleted += OnMidiOutputDeviceEnumerationCompleted;

            watcher.Start();

            // this can be used to unsubsribe, when the containing class gets disposed.
            compositeDisposable.Add(Disposable.Create(() =>
            {
                watcher.Stop();
                watcher.Added -= OnMidiInputDeviceAdded;
                watcher.Removed -= OnMidiInputDeviceRemoved;
                watcher.Updated -= OnMidiInputDeviceUpdated;
                watcher.EnumerationCompleted -= OnMidiInputDeviceEnumerationCompleted;
            }));
    }
```

I guess that NAudio should provide an abstraction over this watcher. On Plattforms not supporting it, you can always enumerate over the list of devices in regular intervals and check for device changes: 

```csharp
            Enumerable.Range(0, MidiIn.NumberOfDevices)
                .Select(id => new MidiDeviceInfo
                {
                    Id = id.ToString(),
                    Name = MidiIn.DeviceInfo(id).ProductName
                })
                .ToList();
```

**Comment by markheath** [comment]
NAudio doesn't provide any abstraction over the UWP MIDI APIs - they work just fine on their own, so I don't see the need to add any additional NAudio support. As you point out, people can always poll the MidiDevices if you want to detect that one has gone away.

**Comment by henon** [comment]
@markheath what if the MIDI device was disconnected only shortly? If I am not mistaken we wouldn't be able to notice this with NAudio and couldn't reconnect to it to ensure a reliable connection.

**Comment by markheath** [comment]
NAudio's classic `MidiIn` (the winmm backend) doesn't surface device add/remove events — winmm gives no clean disconnection signal, so you'd only find out when an operation fails. That's the limitation you hit with the existing example.

The good news is that NAudio 3 adds a WinRT MIDI backend (`WinRTMidiIn` / `WinRTMidiOut` in `NAudio.Wasapi`), and on Windows 10+ that's the right foundation for what you want. `WinRTMidiIn` enumerates devices via `Windows.Devices.Enumeration`, and you can watch that same device collection with a `DeviceWatcher` to get disconnection and reconnection notifications:

```csharp
using Windows.Devices.Enumeration;
using Windows.Devices.Midi;

var selector = MidiInPort.GetDeviceSelector();
var watcher = DeviceInformation.CreateWatcher(selector);

watcher.Removed += (s, update) =>
{
    // update.Id was unplugged — alert the user, dispose your WinRTMidiIn
};
watcher.Added += async (s, info) =>
{
    // a MIDI input appeared (or came back) — reconnect:
    var midiIn = await WinRTMidiIn.CreateAsync(info.Id);
    midiIn.MessageReceived += /* ... */;
    midiIn.Start();
};
watcher.Start();
```

`Removed` gives you the disconnection warning; `Added` lets you auto-reconnect by re-opening the port via `WinRTMidiIn.CreateAsync(deviceId)`. This isn't really achievable with the winmm `MidiIn`, so if device hot-plug matters to you, prefer the WinRT backend.

`WinRTMidiIn` doesn't currently re-surface these as its own `Disconnected`/`Connected` events — you wire the `DeviceWatcher` up yourself for now. It's something we could potentially consider adding in the future if enough people want it.
